### ISSUE TITLE
FromForm Parse Error Reason

### ISSUE THREAD

**Issue by PSeitz** [SergioBenitez/Rocket#670]
Hi,

I wanted to return the reason why an error occured in a FromForm to QueryParameter struct conversion to the user.

Currently only BadParse is returned and the reason is printed to stdout
https://github.com/SergioBenitez/Rocket/blob/df7111143e466c18d1f56377a8d9530a5a306aba/core/codegen/src/decorators/derive_form.rs#L278-L283

Another issue is that the println! maybe could be a target for DOS attack (https://llogiq.github.io/2017/06/01/perf-pitfalls.html). A lot of invalid requests, all waiting for a lock on stdout.
Related: https://github.com/SergioBenitez/Rocket/issues/21

**Comment by SergioBenitez** [comment]
We really shouldn't be doing a `println!` there, but without #21, we don't have much recourse.

As far as a better error goes: it seems to me that the only better thing to do here would be to have a custom error enum for form parse failures. Something like:

```rust
enum FormError<'r> {
    /// The field `.0` was not part of the form to be parsed.
    UnknownField(&'r RawStr),
    /// The field named `.0` failed to parse for the form value `.1`.
    BadFormValue(&'r RawStr, &'r RawStr),
}
```

Is that along the lines of what you'd like?

**Comment by PSeitz** [comment]
Yes, that looks good. 

I'm not sure if this is currently possible, but if we have optional and required fields, this could also be MissingField.
Related: https://github.com/SergioBenitez/Rocket/issues/608

**Comment by SergioBenitez** [comment]
As a note: we can get rid of the `println!` and emit a proper `error!` once #571 lands.
